### PR TITLE
SKUs SDK: regenerate request id on submit conflict from server

### PR DESCRIPTION
**Pull request by evq**
Resolves https://github.com/brave/brave-browser/issues/35742

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by husobee — APPROVED** [review]

File: components/skus/browser/rs/lib/src/storage/kv.rs [hunk 1]
```diff
@@ -331,6 +331,32 @@ where
...
+            state.credentials = Some(CredentialsState::default());
+        }
+
+        if let Some(credentials) = state.credentials.as_mut() {
+            // check and see if we already have this item initialized
+            if let Some(item_credentials) = credentials.items.get_mut(item_id) {
+                // overwrite the request_id with the new value
+                if let Credentials::TimeLimitedV2(item_credentials) = item_credentials {
+                    item_credentials.request_id = Some(request_id.to_string());
+                }
```
Comment (rillian) [inline components/skus/browser/rs/lib/src/storage/kv.rs:353]: Wouldn't it be better to return an error if this is called with something besides as `Credentials::TimeLimitedV2` similar to `get_time_limited_v2_creds`?

**Review by rillian — APPROVED** [review]

File: components/skus/browser/rs/lib/src/storage/kv.rs (hunk 1 above)
Comment (evq) [reply components/skus/browser/rs/lib/src/storage/kv.rs:353]: that makes sense to me

**Comment by evq** [comment]
Completed verification.

Test steps:
1. Start nightly 1.64.21 with dev/staging flags
2. Purchase premium Leo on [account.bravesoftware.com](http://account.bravesoftware.com/)
3. Ensure credentials are retrieved
4. Stop browser
5. Roll timestamp on all creds back in Local State, write down original request id
6. Switch to build from https://github.com/brave/brave-core/pull/21887, start browser
7. Attempt to use Leo, it should show as expired since backround refresh will fail with ( normal conflict )
8. Check SKU state, verify in GeneratedCredentials state and write down request id
9. Restart browser, repeat 7 & 8, verify 8 matches, shut down browser
10. Modify Local State, overwrite request id with the original request id recorded in 5
11. Modify staging backend db, change credential dates to the past
12. Start browser
13. Attempt to use Leo, automatic refresh should succeed
14. Check SKU state, verify new request id was generated ( matching neither that written down in 5 or 8 )

**Comment by stephendonner** [comment]
## Verified `PASSED` using

```


```

### Followed the steps from https://github.com/brave/brave-core/pull/21887#issuecomment-1930744451

**Pull request merged**